Support json formatted strings and Hash/Array merge lookups

abooitt opened the pull request.

Commit:
Support json formatted strings and Hash/Array merge lookups

jsok commented:
I like this idea, my initial comment would be that I think this behaviour should be configurable.
JSON data in vault may or may not be intended to be consumed by hiera.

abooitt commented:
Thanks! I will add configurability.
And what do you think about the addition to use the :hierarchy: list on top of the vault mounts list?

Commit:
Make json parsing optional

Notice that json parsing is only done on :default_field values.

abooitt commented:
Made it optional. And edited README.md accordingly.
Notice that json parsing is currently only attempted on the :default_field values.
Looking forward to your comments.

jsok commented on README.md:
```diff
@@ -48,9 +48,30 @@ To do this, set:
+
+### json parsing of single values - optional
+Only applicable when :default_field is used.
+To use json parsing, set, for example:
```
You need an empty new line before the indented code snippet or it doesn't render correctly.

abooitt replied:
Fixed in next commit.

jsok commented on lib/hiera/backend/vault_backend.rb:
```diff
@@ -61,13 +82,21 @@ def lookup_generic(key, scope)
 
           Hiera.debug("[hiera-vault] Read secret: #{key}")
-          if @config[:default_field]
+          if @config[:default_field] and secret.data.has_key?(@config[:default_field].to_sym) and secret.data.length == 1
```
Why is the `length ==`1 check necessary?

e.g. if `:default_field = "value"` and you have `vault write secret/key value=x other=y else=z`, then the `other` and `else` fields should just be ignored rather.

abooitt replied:
Ignoring data that is in vault does not make sense to me. Ignoring the additional fields makes it impossible to retrieve the contents of those fields with hiera when using a default_field. I'd leave it to the consumer to decide whether to ignore the additional fields.
But I'll make the non-ignoring behavior optional as well, to make sure default behavior stays backwards compatible.

jsok replied:
Point taken, the example I'm imaging is setup like `:default_field = "hiera_only_value"`.

jsok commented on README.md:
```diff
@@ -81,6 +102,8 @@ into their own mount. This could be achieved with a configuration like:
                 - secret
 
 
+The :hierarchy: entries from the hiera configuration are used for each mount.
```
Put the `:hierarchy:` keyword in backticks please.

abooitt replied:
Fixed in next commit.

jsok commented on README.md:
```diff
@@ -48,9 +48,30 @@ To do this, set:
+
+### Lookup type behavior
+
+In case Array or Hash lookup is done, usual array or hash merging takes place based on the configured merge_behavior setting.
```
`merge_behavior` in backticks here as well

abooitt replied:
Fixed

jsok commented on README.md:
```diff
@@ -48,9 +48,30 @@ To do this, set:
+The hiera lookup for `foo` will return just "bar" as a string, but only if it is the only field.
+If `foo` contains more fields, a Hash will be returned, just like with the default behaviour.
+
+### json parsing of single values - optional
```
Since this option depends on `default_field`, can this be a sub-section instead, i.e. `####` rather than `###`.

Also, JSON in uppercase.

abooitt replied:
Fixed both.

Commit:
optionally disable ignoring additional fields

and corrections to README

jsok commented:
OK, I'm happy with the new functionality.

The double negative of `:no_ignore_additional` is hard to follow.
What do you think about options/UI (which takes a bit of inspiration from the `merge_behavior` option of hiera itself):
- `:default_field` - remains the same
- `:default_field_parse` - either `string` (default) or `json`
- `:default_field_behaviour` - either `ignore` or `include` (default) -- basically does the same thing a `::no_ignore_additional`

Commit:
Refactor new config variables

- :parse_json => :default_field_parse: with 'string' or 'json' as valid
  values, default 'string'
- :no_ignore_additional => :default_field_behavior: with 'ignore' or
  'only' as valid values, default 'ignore', which exactly mimics previous
  behavior

abooitt commented:
Refactored the new config variables. The compound condition is now somewhat more readable. :)
And now, default behavior is exactly the same as before: if the default field is not present, nil is returned.

abooitt commented:
Thanks for your attention so far. I appreciate that.
Wondering what your thoughts are about the latest commit, have you seen that I also added the behavior to loop through the hierarchy sources on top of the mounts?

While working further on this, I realized that it might be overkill to use the hierarchy list for lookups in vault. A similar behavior can be configured using the `:mounts` setting within the `:vault` config. 

For example, we have a list of 16 items in our hierarchy, which means for each hiera call within the puppet code 16 connections to vault will be created. And while we have quite some hiera lookups in each catalog compile, this easily results in hitting the open files limit in the vault service and, after increasing the open files limit, in source port exhaustion, in case vault is running on the puppetmaster and listens on localhost only.

As such I will make traversing the global hierarchy sources optional, as well.

Additionally I have been working on creating specific `hiera_vault`, `hiera_vault_array`, and `hiera_vault_hash` functions with corresponding configuration options, to be able to not use the vault backend for normal hiera calls, but only when using the specific `hiera_vault*` functions.
And with these functions it is then also possible to tell them to look in vault only.
For a sneak preview on this you can take a look in the `Flagged` branch in my fork.

Looking forward to your response, while continuing on the stuff.

Commit:
Make using hierarchy optional

jsok commented:
Thanks Arnoud, I'm away on holidays at the moment so I won't be able to get around to this until mid next week. Sorry for the delay!

abooitt commented:
Thanks for letting me know, Jonathan.
Enjoy your holidays.

abooitt commented:
Hi Jonathan, just wondering when you will have some time.

jsok commented on lib/hiera/backend/vault_backend.rb:
```diff
@@ -10,6 +10,21 @@ def initialize()
         @config = Config[:vault]
         @config[:mounts] ||= {}
         @config[:mounts][:generic] ||= ['secret']
+        @config[:use_hierarchy] ||= 'no'
```
I think we should use `true`/`false` here rather than yes/no.

jsok commented on README.md:
```diff
@@ -52,6 +52,59 @@ For example:
+was not specified.
+
+#### JSON parsing of single values - optional
+Only applicable when :default_field is used.
```
s/:default_field/`:default_field`/

jsok commented:
I'm struggling to follow the `use_hierarchy` option's purpose. Can you give a more practical example of when it would be useful?

Can you consider removing it from this PR and opening in another? I'd love to get the JSON support out for you asap.

abooitt commented:
Thank for your comments! I'll handle them.

I will take out the `use_hierarchy` functionality and put it in a new PR.

Commit:
Removed `:use_hierarchy` functionality

and small correction to README

Commit:
Remove superfluous break

which I forgot to remove while removing the `:use_hierarchy`
feature.

jsok commented:
LGTM, will merge now!
